File: utils/fileCleanup.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { v2 as cloudinary } from 'cloudinary';
import dotenv from 'dotenv';

dotenv.config();

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Cloudinary Configuration
cloudinary.config({
    cloud_name: process.env.CLOUDINARY_CLOUD_NAME,
    api_key: process.env.CLOUDINARY_API_KEY,
    api_secret: process.env.CLOUDINARY_API_SECRET
});

/**
 * Extracts all unique /uploads/ or cloudinary paths from a nested object or array.
 */
export const extractUploadPaths = (obj, paths = new Set()) => {
    if (!obj) return paths;

    if (typeof obj === 'string') {
        // Look for URLs containing /uploads/ (Local)
        if (obj.includes('/uploads/')) {
            const fileName = obj.split('/uploads/').pop();
            paths.add({ type: 'local', value: fileName });
        }
        // Look for GridFS URLs
        else if (obj.includes('/api/upload/')) {
            const fileId = obj.split('/api/upload/').pop();
            paths.add({ type: 'gridfs', value: fileId });
        }
        // Look for Cloudinary URLs
        else if (obj.includes('res.cloudinary.com')) {
            // Extract public_id from Cloudinary URL: ...upload/v12345/portfolio_uploads/public_id.jpg
            // We want everything after the version (v[digits]) excluding the extension
            const parts = obj.split('/upload/');
            if (parts.length > 1) {
                const afterUpload = parts[1].split('/');
                // Remove version (v12345) if present
                const pathParts = afterUpload[0].startsWith('v') && !isNaN(afterUpload[0].substring(1)) 
                    ? afterUpload.slice(1) 
                    : afterUpload;
                
                // Rejoin and remove extension
                const fullPath = pathParts.join('/');
                const publicId = fullPath.substring(0, fullPath.lastIndexOf('.')) || fullPath;
                paths.add({ type: 'cloudinary', value: publicId });
            }
        }
    } else if (Array.isArray(obj)) {
        obj.forEach(item => extractUploadPaths(item, paths));
    } else if (typeof obj === 'object') {
        Object.values(obj).forEach(val => extractUploadPaths(val, paths));
    }
    return paths;
};

/**
 * Deletes a file from the appropriate storage (local, GridFS, or Cloudinary).
 */
export const deleteUploadFile = async (item) => {
    if (!item || !item.value) return;
    
    const { type, value } = item;

    if (type === 'local') {
        const safeName = path.basename(value);
        const filePath = path.join(process.cwd(), 'uploads', safeName);
        if (fs.existsSync(filePath)) {
            try {
                fs.unlinkSync(filePath);
                console.log(`🗑️ Deleted local file: ${safeName}`);
            } catch (err) {
                console.error(`❌ Failed to delete local file ${safeName}:`, err.message);
            }
        }
    } 
    else if (type === 'cloudinary') {
        try {
            const result = await cloudinary.uploader.destroy(value);
            console.log(`🗑️ Cloudinary item deleted (${value}):`, result.result);
        } catch (err) {
            console.error(`❌ Cloudinary delete failed for ${value}:`, err.message);
        }
    }
    // GridFS deletion would require bucket access here, which might be overkill for this utility 
    // unless strictly needed. We'll stick to Local and Cloudinary for now.
};

File: app/utils/file_cleanup.py
import re
from typing import Set, Dict, Any, Union, List
import cloudinary
import cloudinary.uploader
import os
import logging

logger = logging.getLogger("portfolio-cleanup")

def extract_upload_paths(obj: Any, paths: Set[tuple] = None) -> Set[tuple]:
    """
    Recursively extracts all unique Cloudinary and local upload paths from any JSON/Object.
    Returns a set of tuples (type, value) to be used with delete_upload_file.
    """
    if paths is None:
        paths = set()
    
    if not obj:
        return paths

    if isinstance(obj, str):
        # 1. Local Uploads
        if "/uploads/" in obj:
            file_name = obj.split("/uploads/").pop()
            paths.add(("local", file_name))
        
        # 2. GridFS / Legacy API uploads
        elif "/api/upload/" in obj:
            file_id = obj.split("/api/upload/").pop()
            paths.add(("gridfs", file_id))
            
        # 3. Cloudinary URLs (Smart extraction)
        elif "res.cloudinary.com" in obj:
            # Matches everything after /upload/ and before the file extension
            # Example: ...upload/v12345/portfolio_uploads/img_99.jpg -> portfolio_uploads/img_99
            parts = obj.split("/upload/")
            if len(parts) > 1:
                after_upload = parts[1].split("/")
                # Skip version tag (v12345) if it exists
                path_parts = after_upload[1:] if after_upload[0].startswith('v') else after_upload
                
                full_path = "/".join(path_parts)
                # Remove file extension (e.g., .jpg, .png)
                public_id = full_path.rsplit('.', 1)[0] if '.' in full_path else full_path
                paths.add(("cloudinary", public_id))
                
    elif isinstance(obj, list):
        for item in obj:
            extract_upload_paths(item, paths)
            
    elif isinstance(obj, dict):
        for val in obj.values():
            extract_upload_paths(val, paths)
            
    return paths

async def delete_upload_file(type: str, value: str):
    """Deletes redundant files from Cloudinary or local storage."""
    if not value:
        return

    if type == "cloudinary":
        try:
            result = cloudinary.uploader.destroy(value)
            logger.info(f"🗑️ Cloudinary item deleted ({value}): {result.get('result')}")
        except Exception as e:
            logger.error(f"❌ Cloudinary delete failed for {value}: {str(e)}")
            
    elif type == "local":
        file_path = os.path.join(os.getcwd(), 'uploads', value)
        if os.path.exists(file_path):
            try:
                os.remove(file_path)
                logger.info(f"🗑️ Deleted local file: {value}")
            except Exception as e:
                logger.error(f"❌ Failed to delete local file {value}: {str(e)}")
